src/runtime/impure/lfortran_intrinsics.c

```diff
@@ -606,8 +606,11 @@ LFORTRAN_API void _lfortran_strcat(char** s1, char** s2, char** dest)
 
 // idx1 and idx2 both start from 1
 LFORTRAN_API char* _lfortran_str_copy(char* s, int32_t idx1, int32_t idx2) {
-    char* dest_char = (char*)malloc(idx2-idx1+2);
-    for (int i=idx1; i <= idx2; i++)
+    if (idx2 == idx1) {
+        idx2 = idx1 + 1;
+    }
+    char* dest_char = (char*)malloc(idx2-idx1+1);
+    for (int i=idx1; i < idx2; i++)
```

Comment:
I think this assumes that `idx1:idx2` is inclusive of both ends if `idx1==idx2`, but exclusive of the right end if `idx1 < idx2`?

That doesn't seem like a good design.

In ASR, we should do inclusive of both ends (if we don't do it for strings, we should). Then here in this functions, it seems easiest to assume inclusive of both ends, I believe that's how the previous logic was.

We might need to "add 1" in AST to ASR for string slicing for the right end, to convert from Python convention to ASR convention.

Reply:
Btw, things will also simplify once we split ArrayRef for string into StringItem (just one index) and StringSection (`idx1`, `idx2`).

Reply:
> I think this assumes that `idx1:idx2` is inclusive of both ends if `idx1==idx2`, but exclusive of the right end if `idx1 < idx2`?

Yes, I think that's how it works in cpython. Here, when `idx1==idx2` we are incrementing idx2, then the loop should be `i < idx2`, isn't that correct? 

in other case i.e, `idx1<idx2` which is string slicing, it should return substring from position `idx1` to `idx2-1`. So the loop should be `i<idx2`.

Cpython example:
```
>>> a = "0123456"
>>> a[2:3]
'2'
```

Reply:
Sure, I'll keep it as per current design.